panic: htlc index mismatch: 1 vs 0

### Background

lnd panicked and process stopped shortly after casual service reload:

```
2018-07-17 11:30:28.694 [INF] CRTR: New channel discovered! Link connects 0235447c7485ff2b945bac5fbc366d54a87389bab8cacf1b64b26ec01e96bd165a and 0371bcbde71d6cc591134e6d95592a9e8baac232a1e975ccd67c111864a5e8d373 with ChannelPoint(837b11a93cf2562738d7ab17560a74cec01b265a30695c2000fc07cc7c50cd46:1): chan_id=557023585794457601, capacity=0.0001 BTC
panic: htlc index mismatch: 1 vs 0

goroutine 95027 [running]:
github.com/lightningnetwork/lnd/lnwallet.(*LightningChannel).restoreStateLogs(0xc423720120, 0xc4223e32c0, 0xc4223e3500, 0xc4223e35c0, 0xc42016c600, 0xc4222064e0, 0x0, 0x0)
        /home/bitcoin/go/src/github.com/lightningnetwork/lnd/lnwallet/channel.go:1807 +0xada
github.com/lightningnetwork/lnd/lnwallet.(*LightningChannel).restoreCommitState(0xc423720120, 0xc42a4f3af8, 0xc42a4f3a78, 0x0, 0xc422e9df80)
        /home/bitcoin/go/src/github.com/lightningnetwork/lnd/lnwallet/channel.go:1687 +0x4a1
github.com/lightningnetwork/lnd/lnwallet.NewLightningChannel(0x170dbe0, 0xc420526a80, 0x7f2e7b8d17a0, 0xc42052a1c0, 0xc42e713500, 0x0, 0x0, 0x17e5410)
        /home/bitcoin/go/src/github.com/lightningnetwork/lnd/lnwallet/channel.go:1380 +0x7ea
main.lndMain.func7(0x5471749bfd1dac89, 0x9ea010e9961c2cd7, 0x316fd78e4bb7c5fe, 0xceb3454beee0b607, 0x0, 0x2, 0xc423b19338)
        /home/bitcoin/go/src/github.com/lightningnetwork/lnd/lnd.go:376 +0x348
main.(*fundingManager).handleFundingLocked(0xc42016c500, 0xc423599760)
        /home/bitcoin/go/src/github.com/lightningnetwork/lnd/fundingmanager.go:2201 +0x455
created by main.(*fundingManager).reservationCoordinator
        /home/bitcoin/go/src/github.com/lightningnetwork/lnd/fundingmanager.go:871 +0x524
```

After the restart, it booted up normally.

### Your environment

* 0.4.2-beta commit=b0288d46773ac6d45e5dc4d5e6a80dd3034d0b9f
* Linux server01 4.9.0-6-amd64 #1 SMP Debian 4.9.88-1+deb9u1 (2018-05-07) x86_64 GNU/Linux
* bitcoind 0.16.1

### Steps to reproduce

I cannot reproduce.

### Expected behaviour

Node should not crash.

### Actual behaviour

Node stopped working, I had to start it manually again.

I saw the same issue on a fresh testnet node

```
2018-09-12 11:26:29.336 [INF] DISC: gossipSyncer(02c9c1c713bfde2b9e5116ced4eb7d26cc8df003f62e0cc2f923212552a311f1b4): applying new update horizon: start=2018-09-12 10:28:38 +0000 UTC, end=2154-10-19 16:56:53 +0000 UTC, backlog_size=334
panic: htlc index mismatch: 1 vs 0

goroutine 53960225 [running]:
github.com/lightningnetwork/lnd/lnwallet.(*LightningChannel).restoreStateLogs(0xc44869c480, 0xc428ccf440, 0xc428ccf500, 0xc428ccf5c0, 0xc434b9b100, 0xc426df4960, 0x0, 0x0)
	/root/go/src/github.com/lightningnetwork/lnd/lnwallet/channel.go:1827 +0xa49
github.com/lightningnetwork/lnd/lnwallet.(*LightningChannel).restoreCommitState(0xc44869c480, 0xc441cefce0, 0xc441cefc60, 0x110cd18, 0x0)
	/root/go/src/github.com/lightningnetwork/lnd/lnwallet/channel.go:1707 +0x49d
github.com/lightningnetwork/lnd/lnwallet.NewLightningChannel(0x11adb60, 0xc420084d80, 0x7f104c421590, 0xc42065a080, 0xc427b78300, 0x1, 0x3b718aa669c71407, 0xc43019a5f0)
	/root/go/src/github.com/lightningnetwork/lnd/lnwallet/channel.go:1400 +0x6a3
main.newServer.func18(0x4287b00f99d0c9b4, 0x4ff3d766cac73d69, 0x9da23efbaec3a98b, 0x7b29cfdaec320fab, 0x0, 0x2, 0xc42216e8c8)
	/root/go/src/github.com/lightningnetwork/lnd/server.go:757 +0x154
main.(*fundingManager).handleFundingLocked(0xc4201e6500, 0xc423749780)
	/root/go/src/github.com/lightningnetwork/lnd/fundingmanager.go:2317 +0x451
created by main.(*fundingManager).reservationCoordinator
	/root/go/src/github.com/lightningnetwork/lnd/fundingmanager.go:892 +0x4f3
```

--- a/discovery/gossiper_test.go
+++ b/discovery/gossiper_test.go
@@ -26,7 +26,6 @@ import (
 	"github.com/lightningnetwork/lnd/chainntnfs"
 	"github.com/lightningnetwork/lnd/channeldb"
 	"github.com/lightningnetwork/lnd/lnpeer"
-	"github.com/lightningnetwork/lnd/lnwallet"
 	"github.com/lightningnetwork/lnd/lnwire"
 	"github.com/lightningnetwork/lnd/routing"
 )
@@ -2315,7 +2314,7 @@ func (p *mockPeer) SendMessage(_ bool, msgs ...lnwire.Message) error {
 
 	return nil
 }
-func (p *mockPeer) AddNewChannel(_ *lnwallet.LightningChannel, _ <-chan struct{}) error {
+func (p *mockPeer) AddNewChannel(_ *channeldb.OpenChannel, _ <-chan struct{}) error {
 	return nil
 }
 func (p *mockPeer) WipeChannel(_ *wire.OutPoint) error { return nil }

--- a/fundingmanager.go
+++ b/fundingmanager.go
@@ -267,7 +267,7 @@ type fundingConfig struct {
 
 	// FindChannel queries the database for the channel with the given
 	// channel ID.
-	FindChannel func(chanID lnwire.ChannelID) (*lnwallet.LightningChannel, error)
+	FindChannel func(chanID lnwire.ChannelID) (*channeldb.OpenChannel, error)
 
 	// TempChanIDSeed is a cryptographically random string of bytes that's
 	// used as a seed to generate pending channel ID's.
@@ -2323,10 +2323,9 @@ func (f *fundingManager) handleFundingLocked(fmsg *fundingLockedMsg) {
 
 	// If the RemoteNextRevocation is non-nil, it means that we have
 	// already processed fundingLocked for this channel, so ignore.
-	if channel.RemoteNextRevocation() != nil {
+	if channel.RemoteNextRevocation != nil {
 		fndgLog.Infof("Received duplicate fundingLocked for "+
 			"ChannelID(%v), ignoring.", chanID)
-		channel.Stop()
 		return
 	}
 
@@ -2334,10 +2333,9 @@ func (f *fundingManager) handleFundingLocked(fmsg *fundingLockedMsg) {
 	// need to create the next commitment state for the remote party. So
 	// we'll insert that into the channel now before passing it along to
 	// other sub-systems.
-	err = channel.InitNextRevocation(fmsg.msg.NextPerCommitmentPoint)
+	err = channel.InsertNextRevocation(fmsg.msg.NextPerCommitmentPoint)
 	if err != nil {
 		fndgLog.Errorf("unable to insert next commitment point: %v", err)
-		channel.Stop()
 		return
 	}
 
@@ -2361,8 +2359,7 @@ func (f *fundingManager) handleFundingLocked(fmsg *fundingLockedMsg) {
 	if err := fmsg.peer.AddNewChannel(channel, f.quit); err != nil {
 		fndgLog.Errorf("Unable to add new channel %v with peer %x: %v",
 			fmsg.peer.IdentityKey().SerializeCompressed(),
-			*channel.ChanPoint, err)
-		channel.Stop()
+			channel.FundingOutpoint, err)
 	}
 }
 

--- a/fundingmanager_test.go
+++ b/fundingmanager_test.go
@@ -178,13 +178,13 @@ func (n *testNode) QuitSignal() <-chan struct{} {
 	return n.shutdownChannel
 }
 
-func (n *testNode) AddNewChannel(channel *lnwallet.LightningChannel,
+func (n *testNode) AddNewChannel(channel *channeldb.OpenChannel,
 	quit <-chan struct{}) error {
 
-	done := make(chan struct{})
+	errChan := make(chan error)
 	msg := &newChannelMsg{
 		channel: channel,
-		done:    done,
+		err:     errChan,
 	}
 
 	select {
@@ -194,12 +194,11 @@ func (n *testNode) AddNewChannel(channel *lnwallet.LightningChannel,
 	}
 
 	select {
-	case <-done:
+	case err := <-errChan:
+		return err
 	case <-quit:
 		return ErrFundingManagerShuttingDown
 	}
-
-	return nil
 }
 
 func init() {
@@ -304,7 +303,8 @@ func createTestFundingManager(t *testing.T, privKey *btcec.PrivateKey,
 			return lnwire.NodeAnnouncement{}, nil
 		},
 		TempChanIDSeed: chanIDSeed,
-		FindChannel: func(chanID lnwire.ChannelID) (*lnwallet.LightningChannel, error) {
+		FindChannel: func(chanID lnwire.ChannelID) (
+			*channeldb.OpenChannel, error) {
 			dbChannels, err := cdb.FetchAllChannels()
 			if err != nil {
 				return nil, err
@@ -312,10 +312,7 @@ func createTestFundingManager(t *testing.T, privKey *btcec.PrivateKey,
 
 			for _, channel := range dbChannels {
 				if chanID.IsChanPoint(&channel.FundingOutpoint) {
-					return lnwallet.NewLightningChannel(
-						signer,
-						nil,
-						channel)
+					return channel, nil
 				}
 			}
 
@@ -992,14 +989,14 @@ func assertHandleFundingLocked(t *testing.T, alice, bob *testNode) {
 	// They should both send the new channel state to their peer.
 	select {
 	case c := <-alice.newChannels:
-		close(c.done)
+		close(c.err)
 	case <-time.After(time.Second * 15):
 		t.Fatalf("alice did not send new channel to peer")
 	}
 
 	select {
 	case c := <-bob.newChannels:
-		close(c.done)
+		close(c.err)
 	case <-time.After(time.Second * 15):
 		t.Fatalf("bob did not send new channel to peer")
 	}

--- a/htlcswitch/link_test.go
+++ b/htlcswitch/link_test.go
@@ -1477,7 +1477,8 @@ func (m *mockPeer) SendMessage(sync bool, msgs ...lnwire.Message) error {
 	}
 	return nil
 }
-func (m *mockPeer) AddNewChannel(_ *lnwallet.LightningChannel, _ <-chan struct{}) error {
+func (m *mockPeer) AddNewChannel(_ *channeldb.OpenChannel,
+	_ <-chan struct{}) error {
 	return nil
 }
 func (m *mockPeer) WipeChannel(*wire.OutPoint) error {

--- a/htlcswitch/mock.go
+++ b/htlcswitch/mock.go
@@ -543,7 +543,7 @@ func (s *mockServer) Address() net.Addr {
 	return nil
 }
 
-func (s *mockServer) AddNewChannel(channel *lnwallet.LightningChannel,
+func (s *mockServer) AddNewChannel(channel *channeldb.OpenChannel,
 	cancel <-chan struct{}) error {
 
 	return nil

--- a/lnpeer/peer.go
+++ b/lnpeer/peer.go
@@ -5,7 +5,7 @@ import (
 
 	"github.com/btcsuite/btcd/btcec"
 	"github.com/btcsuite/btcd/wire"
-	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/channeldb"
 	"github.com/lightningnetwork/lnd/lnwire"
 )
 
@@ -19,7 +19,7 @@ type Peer interface {
 
 	// AddNewChannel adds a new channel to the peer. The channel should fail
 	// to be added if the cancel channel is closed.
-	AddNewChannel(channel *lnwallet.LightningChannel, cancel <-chan struct{}) error
+	AddNewChannel(channel *channeldb.OpenChannel, cancel <-chan struct{}) error
 
 	// WipeChannel removes the channel uniquely identified by its channel
 	// point from all indexes associated with the peer.

--- a/peer.go
+++ b/peer.go
@@ -59,11 +59,11 @@ type outgoingMsg struct {
 	errChan chan error // MUST be buffered.
 }
 
-// newChannelMsg packages an lnwallet.LightningChannel with a channel that
-// allows the receiver of the request to report when the funding transaction
-// has been confirmed and the channel creation process completed.
+// newChannelMsg packages a channeldb.OpenChannel with a channel that allows
+// the receiver of the request to report when the funding transaction has been
+// confirmed and the channel creation process completed.
 type newChannelMsg struct {
-	channel *lnwallet.LightningChannel
+	channel *channeldb.OpenChannel
 	err     chan error
 }
 
@@ -1522,9 +1522,9 @@ out:
 		// funding workflow. We'll initialize the necessary local
 		// state, and notify the htlc switch of a new link.
 		case newChanReq := <-p.newChannels:
-			chanPoint := newChanReq.channel.ChannelPoint()
-			chanID := lnwire.NewChanIDFromOutPoint(chanPoint)
 			newChan := newChanReq.channel
+			chanPoint := &newChan.FundingOutpoint
+			chanID := lnwire.NewChanIDFromOutPoint(chanPoint)
 
 			// Make sure this channel is not already active.
 			p.activeChanMtx.Lock()
@@ -1534,7 +1534,6 @@ out:
 
 				p.activeChanMtx.Unlock()
 				close(newChanReq.err)
-				newChanReq.channel.Stop()
 
 				// If we're being sent a new channel, and our
 				// existing channel doesn't have the next
@@ -1548,7 +1547,7 @@ out:
 					"FundingLocked for ChannelPoint(%v)",
 					chanPoint)
 
-				nextRevoke := newChan.RemoteNextRevocation()
+				nextRevoke := newChan.RemoteNextRevocation
 				err := currentChan.InitNextRevocation(nextRevoke)
 				if err != nil {
 					peerLog.Errorf("unable to init chan "+
@@ -1562,7 +1561,21 @@ out:
 			// If not already active, we'll add this channel to the
 			// set of active channels, so we can look it up later
 			// easily according to its channel ID.
-			p.activeChannels[chanID] = newChan
+			lnChan, err := lnwallet.NewLightningChannel(
+				p.server.cc.signer, p.server.witnessBeacon,
+				newChan,
+			)
+			if err != nil {
+				p.activeChanMtx.Unlock()
+				err := fmt.Errorf("unable to create "+
+					"LightningChannel: %v", err)
+				peerLog.Errorf(err.Error())
+
+				newChanReq.err <- err
+				continue
+			}
+
+			p.activeChannels[chanID] = lnChan
 			p.activeChanMtx.Unlock()
 
 			peerLog.Infof("New channel active ChannelPoint(%v) "+
@@ -1578,6 +1591,7 @@ out:
 					"block: %v", err)
 				peerLog.Errorf(err.Error())
 
+				lnChan.Stop()
 				newChanReq.err <- err
 				continue
 			}
@@ -1589,6 +1603,7 @@ out:
 					"chain events: %v", err)
 				peerLog.Errorf(err.Error())
 
+				lnChan.Stop()
 				newChanReq.err <- err
 				continue
 			}
@@ -1598,7 +1613,7 @@ out:
 			// forwarded. For fees we'll use the default values, as
 			// they currently are always set to the default values
 			// at initial channel creation.
-			fwdMinHtlc := newChan.FwdMinHtlc()
+			fwdMinHtlc := lnChan.FwdMinHtlc()
 			defaultPolicy := p.server.cc.routingPolicy
 			forwardingPolicy := &htlcswitch.ForwardingPolicy{
 				MinHTLC:       fwdMinHtlc,
@@ -1609,7 +1624,7 @@ out:
 
 			// Create the link and add it to the switch.
 			err = p.addLink(
-				chanPoint, newChan, forwardingPolicy,
+				chanPoint, lnChan, forwardingPolicy,
 				chainEvents, currentHeight, false,
 			)
 			if err != nil {
@@ -1618,6 +1633,7 @@ out:
 					p.PubKey())
 				peerLog.Errorf(err.Error())
 
+				lnChan.Stop()
 				newChanReq.err <- err
 				continue
 			}
@@ -2182,7 +2198,7 @@ func (p *peer) Address() net.Addr {
 // added if the cancel channel is closed.
 //
 // NOTE: Part of the lnpeer.Peer interface.
-func (p *peer) AddNewChannel(channel *lnwallet.LightningChannel,
+func (p *peer) AddNewChannel(channel *channeldb.OpenChannel,
 	cancel <-chan struct{}) error {
 
 	errChan := make(chan error, 1)

--- a/peer.go
+++ b/peer.go
@@ -64,7 +64,7 @@ type outgoingMsg struct {
 // has been confirmed and the channel creation process completed.
 type newChannelMsg struct {
 	channel *lnwallet.LightningChannel
-	done    chan struct{}
+	err     chan error
 }
 
 // closeMsgs is a wrapper struct around any wire messages that deal with the
@@ -1533,7 +1533,7 @@ out:
 					"ignoring.", chanPoint)
 
 				p.activeChanMtx.Unlock()
-				close(newChanReq.done)
+				close(newChanReq.err)
 				newChanReq.channel.Stop()
 
 				// If we're being sent a new channel, and our
@@ -1574,15 +1574,22 @@ out:
 			// TODO(roasbeef): panic on below?
 			_, currentHeight, err := p.server.cc.chainIO.GetBestBlock()
 			if err != nil {
-				peerLog.Errorf("unable to get best block: %v", err)
+				err := fmt.Errorf("unable to get best "+
+					"block: %v", err)
+				peerLog.Errorf(err.Error())
+
+				newChanReq.err <- err
 				continue
 			}
 			chainEvents, err := p.server.chainArb.SubscribeChannelEvents(
 				*chanPoint,
 			)
 			if err != nil {
-				peerLog.Errorf("unable to subscribe to chain "+
-					"events: %v", err)
+				err := fmt.Errorf("unable to subscribe to "+
+					"chain events: %v", err)
+				peerLog.Errorf(err.Error())
+
+				newChanReq.err <- err
 				continue
 			}
 
@@ -1606,12 +1613,16 @@ out:
 				chainEvents, currentHeight, false,
 			)
 			if err != nil {
-				peerLog.Errorf("can't register new channel "+
+				err := fmt.Errorf("can't register new channel "+
 					"link(%v) with NodeKey(%x)", chanPoint,
 					p.PubKey())
+				peerLog.Errorf(err.Error())
+
+				newChanReq.err <- err
+				continue
 			}
 
-			close(newChanReq.done)
+			close(newChanReq.err)
 
 		// We've just received a local request to close an active
 		// channel. If will either kick of a cooperative channel
@@ -2174,10 +2185,10 @@ func (p *peer) Address() net.Addr {
 func (p *peer) AddNewChannel(channel *lnwallet.LightningChannel,
 	cancel <-chan struct{}) error {
 
-	newChanDone := make(chan struct{})
+	errChan := make(chan error, 1)
 	newChanMsg := &newChannelMsg{
 		channel: channel,
-		done:    newChanDone,
+		err:     errChan,
 	}
 
 	select {
@@ -2191,12 +2202,11 @@ func (p *peer) AddNewChannel(channel *lnwallet.LightningChannel,
 	// We pause here to wait for the peer to recognize the new channel
 	// before we close the channel barrier corresponding to the channel.
 	select {
-	case <-newChanDone:
+	case err := <-errChan:
+		return err
 	case <-p.quit:
 		return ErrPeerExiting
 	}
-
-	return nil
 }
 
 // StartTime returns the time at which the connection was established if the

--- a/server.go
+++ b/server.go
@@ -743,7 +743,9 @@ func newServer(listenAddrs []net.Addr, chanDB *channeldb.DB, cc *chainControl,
 		},
 		NotifyWhenOnline: s.NotifyWhenOnline,
 		TempChanIDSeed:   chanIDSeed,
-		FindChannel: func(chanID lnwire.ChannelID) (*lnwallet.LightningChannel, error) {
+		FindChannel: func(chanID lnwire.ChannelID) (
+			*channeldb.OpenChannel, error) {
+
 			dbChannels, err := chanDB.FetchAllChannels()
 			if err != nil {
 				return nil, err
@@ -751,10 +753,7 @@ func newServer(listenAddrs []net.Addr, chanDB *channeldb.DB, cc *chainControl,
 
 			for _, channel := range dbChannels {
 				if chanID.IsChanPoint(&channel.FundingOutpoint) {
-					return lnwallet.NewLightningChannel(
-						cc.signer, s.witnessBeacon,
-						channel,
-					)
+					return channel, nil
 				}
 			}
 

